src/AppInstallerCLICore/Workflows/PortableInstallFlow.cpp

```diff
@@ -0,0 +1,307 @@
+            }
+
+            if (appsAndFeaturesEntry.DisplayName.empty())
+            {
+                appsAndFeaturesEntry.DisplayName = displayName;
+            }
+            if (appsAndFeaturesEntry.DisplayVersion.empty())
+            {
+                appsAndFeaturesEntry.DisplayVersion = displayVersion;
+            }
+            if (appsAndFeaturesEntry.Publisher.empty())
+            {
+                appsAndFeaturesEntry.Publisher = publisher;
+            }
+            if (appsAndFeaturesEntry.ProductCode.empty())
+            {
+                appsAndFeaturesEntry.ProductCode = packageId;
+            }
+
+            return appsAndFeaturesEntry;
+        }
+
+        void AppendExeExtension(std::string& value)
+        {
+            if (!HasSuffix(value, ".exe"))
+            {
+                value += ".exe";
+            }
+        }
+
+        bool AddPortableLinksDirToPathRegistry(Manifest::ScopeEnum& scope, Utility::Architecture& arch, bool& isPathModified)
+        {
+            Key key;
+            if (scope == Manifest::ScopeEnum::Machine)
+            {
+                key = Registry::Key::CreateKeyAndOpen(HKEY_LOCAL_MACHINE, Normalize(s_PathSubkey_Machine));
+            }
+            else
+            {
+                key = Registry::Key::CreateKeyAndOpen(HKEY_CURRENT_USER, Normalize(s_PathSubkey_User));
+            }
+
+            std::string portableLinksDir = GetPortableLinksLocation(scope, arch).u8string();
+            AICLI_LOG(CLI, Info, << "Adding to Path environment variable: " << portableLinksDir);
+            std::wstring pathKey = { L"Path" };
+            std::string pathValue = key[pathKey]->GetValue<Value::Type::String>();
+            if (pathValue.find(portableLinksDir) == std::string::npos)
+            {
+                std::string modifiedPathValue = pathValue;
+
+                if (modifiedPathValue.back() != ';')
+                {
+                    modifiedPathValue += ";";
+                }
+
+                modifiedPathValue += portableLinksDir + ";";
+                bool result = key.SetKeyValue(pathKey, ConvertToUTF16(modifiedPathValue), REG_EXPAND_SZ);
+                if (result)
+                {
+                    isPathModified = true;
+                }
+            }
+
+            return true;
+        }
+
+        bool AddPortableEntryToUninstallRegistry(Manifest::ScopeEnum& scope, std::string_view packageId, AppsAndFeaturesEntry& entry)
+        {
+            HKEY root = (scope == Manifest::ScopeEnum::Machine) ? HKEY_LOCAL_MACHINE : HKEY_CURRENT_USER;
+            std::wstring productCode = ConvertToUTF16(entry.ProductCode);
+            std::wstring displayName = ConvertToUTF16(entry.DisplayName);
+            std::wstring displayVersion = ConvertToUTF16(entry.DisplayVersion);
+            std::wstring publisher = ConvertToUTF16(entry.Publisher);
+            std::wstring uninstallString = L"winget uninstall --id " + ConvertToUTF16(packageId);

```

Comment:
Is the package id necessary here, or could the product code from the passed AppsAndFeaturesEntry be used instead, since the entry will come from the `GetAppsAndFeaturesEntryForPortableInstall` function ?

Reply:
This sparked quite a long conversation about uninstall, and the ideas that:
1. Uninstall should be complete
2. Uninstall should be resilient

With completeness being somewhat obvious, but meaning that anything that was written by winget will be removed.

Resilience means preserving the completeness principle in the face of an incomplete install (power loss) or the DesktopAppInstaller package being reset (and thus losing any data stored in it's state location).

Ultimately, this means that you are correct, the uninstall string should not be based on an identifier that exists outside of the ARP entry.  We might want to consider adding an argument specifically for this, aka `winget uninstall --product-code foo` so that we don't have to worry as much about random conflicts (in addition to the ones between the `Uninstall` keys that we are currently ignoring...).